File: examples/Hedgehog.Xunit.Examples.CSharp/GenAttributeVsContainer.cs
namespace Hedgehog.Xunit.Examples.CSharp;

using Hedgehog;
using Hedgehog.Linq;
using Hedgehog.Xunit;
using Gen = Linq.Gen;
using Range = Linq.Range;

public class PositiveAndNegativeWithAutoGenConfigContainer
{
    public record PositiveInt(int Value);
    public record NegativeInt(int Value);

    public class Generators
    {
        public static Gen<PositiveInt> GenPositiveInt =>
            from i in Gen.Int32(Range.Constant(1, int.MaxValue))
            select new PositiveInt(i);

        public static Gen<NegativeInt> GenNegativeInt =>
            from i in Gen.Int32(Range.Constant(int.MinValue, -1))
            select new NegativeInt(i);

        public static AutoGenConfig _ => GenX.defaults
            .WithGenerator(GenPositiveInt)
            .WithGenerator(GenNegativeInt);
    }

    [Property(typeof(Generators))]
    public bool ResultOfAddingPositiveAndNegativeLessThanPositive(
        PositiveInt positive,
        NegativeInt negative) =>
            positive.Value + negative.Value < positive.Value;
}

public class PositiveAndNegativeWithGenAttribute
{
    public class Negative : GenAttribute<int>
    {
        public override Gen<int> Generator => Gen.Int32(Range.Constant(int.MinValue, -1));
    }

    public class Positive : GenAttribute<int>
    {
        public override Gen<int> Generator => Gen.Int32(Range.Constant(1, int.MaxValue));
    }

    [Property]
    public bool ResultOfAddingPositiveAndNegativeLessThanPositive(
        [Positive] int positive,
        [Negative] int negative) =>
            positive + negative < positive;
}

public class PositiveAndNegativeWithParameterizedGenAttribute
{
    public class Int32Range : GenAttribute<int>
    {
        private readonly int _min;
        private readonly int _max;

        public Int32Range(int min, int max)
        {
            _min = min;
            _max = max;
        }

        public override Gen<int> Generator => Gen.Int32(Range.Constant(_min, _max));
    }

    [Property]
    public bool ResultOfAddingPositiveAndNegativeLessThanPositive(
        [Int32Range(1, int.MaxValue)] int positive,
        [Int32Range(int.MinValue, -1)] int negative) =>
            positive + negative < positive;
}

File: examples/Hedgehog.Xunit.Examples.FSharp/GenAttributeVsContainer.fs
module GenAttributeVsContainer

open System
open Hedgehog
open Hedgehog.Xunit

let positiveInt() = Range.constant 1 Int32.MaxValue |> Gen.int32 
let negativeInt() = Range.constant Int32.MinValue 1 |> Gen.int32

module ``With AutoGenConfig Container`` =
  type PositiveInt = { Value: int }
  type NegativeInt = { Value: int }

  type AutoGenConfigContainer =
    static member __ =
      GenX.defaults
        |> AutoGenConfig.addGenerator (positiveInt() |> Gen.map(fun x -> { PositiveInt.Value = x }))
        |> AutoGenConfig.addGenerator (negativeInt() |> Gen.map(fun x -> { NegativeInt.Value = x }))

  [<Property(typeof<AutoGenConfigContainer>)>]
  let ``Positive + Negative < Positive`` (positive: PositiveInt) (negative: NegativeInt) =
    positive.Value + negative.Value < positive.Value

module ``With GenAttribute`` =
  type PosInt() =
    inherit GenAttribute<int>()
    override _.Generator = positiveInt()

  type NegInt() =
    inherit GenAttribute<int>()
    override _.Generator = negativeInt()

  [<Property>]
  let ``Positive + Negative < Positive`` ([<PosInt>] positive) ([<NegInt>] negative) =
    positive + negative < positive

module ``With Parameterized GenAttribute`` =
  type IntRange(min: int32, max: int32) =
    inherit GenAttribute<int>()
    override _.Generator = Range.constant min max |> Gen.int32

  [<Property>]
  let ``Positive + Negative < Positive, parameterized``
    ([<IntRange(0, Int32.MaxValue)>] positive)
    ([<IntRange(Int32.MinValue, 0)>] negative) =
    positive + negative < positive
